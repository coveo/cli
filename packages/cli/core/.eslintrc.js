const base = require('@coveo/cli-commons-dev/.eslintrc');
const config = {...base};
config.parserOptions = {...config.parserOptions, project: 'tsconfig.json'};

module.exports = config;
