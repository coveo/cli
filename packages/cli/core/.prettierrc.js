const base = require('@coveo/cli-commons-dev/.prettierrc');

module.exports = {
  ...base,
};
