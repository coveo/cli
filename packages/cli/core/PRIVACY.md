# Privacy

Coveo collects usage data and metrics (telemetry) to help improve Coveo CLI.

We collect information related to the use of the CLI and plug-ins, such as which commands and parameters were run, performance and error data.

We use these data to improve the Coveo platform by looking at trends in command executions, to better understand user journey, and to reduce friction in the user experience.

We also use error data to improve the CLI.

You can decide to opt in or opt out at any moment by using the config:set command.
