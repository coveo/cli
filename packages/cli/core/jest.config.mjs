import {dirname, resolve} from 'path';
import {fileURLToPath} from 'url';
import {pathsToModuleNameMapper} from 'ts-jest';
// In the following statement, replace `./tsconfig` with the path to your `tsconfig` file
// which contains the path mapping (ie the `compilerOptions.paths` option):
import tsconfig from './tsconfig.json' assert {type: 'json'};

process.env.TS_NODE_PROJECT = resolve(
  dirname(fileURLToPath(import.meta.url)),
  './tsconfig.dev.json'
);
/*
 * For a detailed explanation regarding each configuration property, visit:
 * https://jestjs.io/docs/configuration
 */

/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
export default {
  roots: ['<rootDir>'],
  modulePaths: ['.'], // <-- This will be set to 'baseUrl' value
  extensionsToTreatAsEsm: ['.ts'],
  preset: 'ts-jest/presets/default-esm',
  moduleNameMapper: pathsToModuleNameMapper(tsconfig.compilerOptions.paths, {
    useESM: true,
    rootDir: '<rootDir>/src',
  }),
  clearMocks: true,
  collectCoverage: true,
  collectCoverageFrom: ['<rootDir>/**/*.{ts,tsx}'],
  coverageDirectory: 'coverage',
  testEnvironment: 'node',
};
