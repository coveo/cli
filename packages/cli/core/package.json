{
  "name": "@coveo/cli",
  "version": "2.0.0",
  "author": "Coveo",
  "description": "Coveo CLI",
  "engines": {
    "node": "16.x || 18.x"
  },
  "dependencies": {
    "@amplitude/identify": "^1.9.0",
    "@amplitude/node": "^1.9.0",
    "@coveo/cli-commons": "2.1.0",
    "@coveo/cli-plugin-source": "2.0.0",
    "@coveo/platform-client": "37.9.0",
    "@coveo/push-api-client": "^2.7.7",
    "@oclif/core": "1.20.4",
    "@oclif/plugin-help": "5.1.19",
    "@oclif/plugin-plugins": "2.1.7",
    "@oclif/plugin-update": "3.0.7",
    "@oclif/plugin-version": "1.1.3",
    "abortcontroller-polyfill": "^1.7.1",
    "archiver": "^5.3.0",
    "async-retry": "^1.3.1",
    "axios": "^1.0.0",
    "chalk": "4.1.2",
    "cli-progress": "^3.9.1",
    "extract-zip": "^2.0.1",
    "fs-extra": "^10.0.0",
    "https-proxy-agent": "^5.0.0",
    "is-unicode-supported": "^1.3.0",
    "isomorphic-fetch": "^3.0.0",
    "json2csv": "^5.0.6",
    "jsonschema": "1.4.1",
    "node-machine-id": "^1.1.12",
    "open": "^8.4.0",
    "semver": "^7.3.4",
    "tmp": "^0.2.1",
    "ts-dedent": "2.2.0",
    "tslib": "2.4.1",
    "undici": "5.12.0"
  },
  "devDependencies": {
    "@amplitude/types": "1.10.2",
    "@babel/core": "7.20.2",
    "@coveo/angular": "1.35.17",
    "@coveo/cli-commons-dev": "5.0.0",
    "@coveo/cra-template": "1.36.0",
    "@coveo/create-atomic": "1.35.15",
    "@coveo/create-headless-vue": "1.0.0",
    "@oclif/test": "2.2.12",
    "@types/archiver": "5.3.1",
    "@types/async-retry": "1.4.5",
    "@types/cli-progress": "3.11.0",
    "@types/fs-extra": "9.0.13",
    "@types/inquirer": "9.0.3",
    "@types/jest": "29.2.3",
    "@types/json2csv": "5.0.3",
    "@types/node": "18.11.9",
    "@types/semver": "7.3.13",
    "@types/tmp": "0.2.3",
    "chai": "4.3.7",
    "eslint": "8.28.0",
    "fancy-test": "2.0.7",
    "jest": "29.3.1",
    "mock-stdin": "1.0.0",
    "oclif": "3.2.28",
    "prettier": "2.8.0",
    "rimraf": "3.0.2",
    "stdout-stderr": "0.1.13",
    "strip-ansi": "6.0.1",
    "ts-jest": "29.0.3",
    "ts-node": "10.9.1",
    "tsconfig-paths": "4.1.0",
    "typescript": "4.9.3"
  },
  "peerDependencies": {
    "@angular/cli": "^15.0.0",
    "netlify-cli": "^12.0.0",
    "create-react-app": "*"
  },
  "peerDependenciesMeta": {
    "@angular/cli": {
      "optional": true
    },
    "netlify-cli": {
      "optional": true
    },
    "create-react-app": {
      "optional": true
    }
  },
  "files": [
    "/bin",
    "/lib",
    "/npm-shrinkwrap.json",
    "/oclif.manifest.json"
  ],
  "homepage": "https://github.com/coveo/cli",
  "keywords": [
    "oclif"
  ],
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "bin": {
    "coveo": "./bin/run"
  },
  "oclif": {
    "repositoryPrefix": "<%- repo %>/blob/v<%- version %>/packages/cli/core/<%- commandPath %>",
    "description": "Coveo CLI",
    "commands": "./lib/commands",
    "bin": "coveo",
    "plugins": [
      "@oclif/plugin-help",
      "@oclif/plugin-update",
      "@oclif/plugin-version",
      "@oclif/plugin-plugins",
      "@coveo/cli-plugin-source"
    ],
    "oclif": {
      "additionalHelpFlags": [
        "-h"
      ],
      "additionalVersionFlags": [
        "-v"
      ]
    },
    "scope": "coveo",
    "prefix": "cli-plugin",
    "topicSeparator": ":",
    "topics": {
      "auth": {
        "description": "manage authentification against the Coveo platform"
      },
      "config": {
        "description": "manage Coveo CLI configuration"
      },
      "org": {
        "description": "manage Coveo organizations"
      },
      "org:resources": {
        "description": "manage your organization resources (e.g., Query Pipelines, Fields or Source Configurations) with Coveo Snapshots"
      },
      "org:resources:model": {
        "description": "manage your Snapshot Pull Model"
      },
      "org:search": {
        "description": "perform queries against a coveo index"
      },
      "ui": {
        "description": "manage user interface deployments"
      },
      "ui:create": {
        "description": "create a user interface powered by different front end frameworks and Coveo Headless"
      }
    },
    "hooks": {
      "init": "./lib/hooks/init/set-global-config",
      "analytics": "./lib/hooks/analytics/analytics",
      "command_not_found": "./lib/hooks/commandNotFound/commandNotFound",
      "prerun": "./lib/hooks/prerun/prerun",
      "postrun": "./lib/hooks/postrun/postrun"
    },
    "macos": {
      "identifier": "com.coveo.cli"
    },
    "update": {
      "s3": {
        "host": "https://static.cloud.coveo.com/cli",
        "bucket": "https://static.cloud.coveo.com/cli"
      }
    }
  },
  "repository": "coveo/cli",
  "scripts": {
    "build": "rimraf lib && tsc -b tsconfig.build.json",
    "test": "jest --colors",
    "lint": "prettier --check . && eslint .",
    "release:phase1": "node --experimental-specifier-resolution=node ../../../scripts/releaseV2/phase1-bump-all-packages.mjs",
    "postpack": "rimraf oclif.manifest.json",
    "prepublishOnly": "rimraf lib && npm run build && oclif manifest && oclif readme && node ./scripts/post-readme.mjs"
  },
  "types": "lib/index.d.ts",
  "publishConfig": {
    "access": "public"
  }
}
