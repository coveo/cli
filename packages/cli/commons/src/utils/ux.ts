import {CliUx} from '@oclif/core';
import {red, green, magenta} from 'chalk';

export function startSpinner(task: string, status?: string) {
  if (CliUx.ux.action.running) {
    CliUx.ux.action.stop(green('✔'));
  }
  CliUx.ux.action.start(task, status);
}

export function stopSpinner(options?: {success?: boolean; message?: string}) {
  if (!CliUx.ux.action.running) {
    return;
  }
  const defaultOptions = {success: true, message: ''};
  const {success, message} = {...defaultOptions, ...options};
  const symbol = success ? green('✔') : red.bold('!');
  CliUx.ux.action.stop(`${symbol} ${message}`.trimEnd());
}

export const formatOrgId = (orgId: TemplateStringsArray | string) =>
  magenta(orgId);
