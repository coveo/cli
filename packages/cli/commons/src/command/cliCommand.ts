import {Command} from '@oclif/core';
import {CLIBaseError} from '../errors/cliBaseError';
import {stopSpinner} from '../utils/ux';
import {wrapError} from '../errors/wrapError';
import {Trackable} from '../preconditions/trackable';

/**
 * A base command to standadize error handling, analytic tracking and logging.
 *
 * @class CLICommand
 * @extends {Command}
 */
export abstract class CLICommand extends Command {
  public abstract run(): PromiseLike<any>;

  /**
   * If you extend or overwrite the catch method in your command class, make sure it returns `return super.catch(err)`
   *
   * @param {*} [err]
   * @see [Oclif Error Handling](https://oclif.io/docs/error_handling)
   */
  @Trackable()
  protected async catch(err?: any): Promise<CLIBaseError | never> {
    // Debug raw error
    this.debug(err);

    // Convert all other errors to CLIBaseErrors for consistency
    const error = wrapError(err);

    // Let oclif handle errors
    throw error;
  }

  protected async finally(arg?: unknown) {
    let error: CLIBaseError | undefined;
    if (arg) {
      error = wrapError(arg);
    }

    try {
      stopSpinner({success: error === undefined});
    } catch {}

    return super.finally(error);
  }

  public get identifier(): string {
    return this.ctor.id;
  }
}
