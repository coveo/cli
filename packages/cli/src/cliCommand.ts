import {CliUx, Command} from '@oclif/core';
import {CLIBaseError, SeverityLevel} from './lib/errors/CLIBaseError';
import {Trackable} from './lib/decorators/preconditions/trackable';
import {stopSpinner} from './lib/utils/ux';
import {wrapError} from './lib/errors/wrapError';
import {noop} from './lib/utils/misc';

export abstract class CliCommand extends Command {
  public abstract run(): PromiseLike<any>;

  @Trackable()
  protected async catch(err?: any): Promise<CLIBaseError | never> {
    // Debug raw error
    this.debug(err);

    // Convert all other errors to CLIBaseErrors for consistency
    let error = wrapError(err);

    return this.isFatalError(error)
      ? this.handleFatalError(error)
      : this.handleNonFatalError(error);
  }

  protected async finally(err: Error | undefined) {
    try {
      stopSpinner(err);
    } catch {
      noop();
    }

    return super.finally(err);
  }

  private isFatalError(error: CLIBaseError): boolean {
    return error.severityLevel === SeverityLevel.Error;
  }

  private handleNonFatalError(error: CLIBaseError): CLIBaseError {
    CliUx.ux.log();
    CliUx.ux[error.severityLevel]('\n' + error.message);
    return error;
  }

  private handleFatalError(error: CLIBaseError): never {
    // Let oclif handle errors
    throw error;
  }
}
