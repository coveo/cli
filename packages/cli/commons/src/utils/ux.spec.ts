import {CliUx} from '@oclif/core';
import {CLICommand} from '../command/cliCommand';
import {formatOrgId, startSpinner, stopSpinner} from './ux';
import {stderr} from 'stdout-stderr';
class FakeCommandNoError extends CLICommand {
  public async run() {
    startSpinner('spinniboiii');
  }
}

class FakeCommandWithNonStandardError extends CLICommand {
  public async run() {
    startSpinner('spinniboiii');
    throw {random: 'object'};
  }
}

class FakeCommandWithError extends CLICommand {
  public async run() {
    startSpinner('spinniboiii');
    throw new Error('no more spinning');
  }
}

describe('ux', () => {
  beforeAll(() => {
    // @oclif/core has an issue preventing the first stream write to be properly mocked.
    // This is a bodge.
    stderr.start();
    CliUx.ux.action.start('test');
    CliUx.ux.action.stop();
    stderr.stop();
  });

  beforeEach(() => {
    stderr.start();
  });

  afterEach(() => {
    stderr.stop();
  });

  describe('startSpinner() & stopSpinner()', () => {
    describe('when spinner is ended with no argument', () => {
      it('no spinner should be running', () => {
        expect(CliUx.ux.action.running).toBe(false);
      });

      it('should start a spinner instance', () => {
        startSpinner('something');
        expect(CliUx.ux.action.running).toBe(true);
        stopSpinner();
        expect(CliUx.ux.action.running).toBe(false);
      });

      it('should stop running task without error', () => {
        startSpinner('starting a process');
        stopSpinner();
        expect(stderr.output).toContain('starting a process... ✔');
      });

      it('should print new task after ending previous task', () => {
        startSpinner('starting the car');
        startSpinner('installing stuff');
        stopSpinner();
        expect(stderr.output).toContain('starting the car... ✔');
        expect(stderr.output).toContain('installing stuff... ✔');
      });
    });

    describe('when spinner is ended with an argument', () => {
      it('should stop running task with the error', () => {
        startSpinner('doing something');
        stopSpinner({success: false});
        expect(stderr.output).toContain('doing something... !');
      });

      it('should stop running task with a success', () => {
        startSpinner('playing a game');
        stopSpinner({success: true});
        expect(stderr.output).toContain('playing a game... ✔');
      });

      describe('when no spinner is running', () => {
        it('should not print anything', () => {
          stopSpinner();
          stopSpinner();
          stopSpinner();
          expect(stderr.output).toBe('');
        });
      });

      describe('when spinning within a command context', () => {
        describe('when the command does not throw errors', () => {
          it('it should end with a ✔', async () => {
            await FakeCommandNoError.run();
            expect(stderr.output).toMatch('spinniboiii... ✔');
          });
        });

        describe('when the command throws any kind of error', () => {
          it.each([
            {command: FakeCommandWithError},
            {command: FakeCommandWithNonStandardError},
          ])('it should end with a !', async ({command}) => {
            try {
              await command.run();
            } catch (error) {
              expect(stderr.output).toMatch('spinniboiii... !');
            }
          });

          describe('when the command throws any kind of error', () => {
            it.each([
              {command: FakeCommandWithError},
              {command: FakeCommandWithNonStandardError},
            ])('it should end with a !', async ({command}) => {
              try {
                stderr.start();
                await command.run();
                stderr.stop();
              } catch (error) {
                expect(stderr.output).toMatch('spinniboiii... !');
              }
            });
          });
        });
      });
    });
  });

  describe('formatOrgId()', () => {
    it('colors the color magenta', () => {
      expect(formatOrgId('myOrgId')).toMatchSnapshot();
    });
  });
});
