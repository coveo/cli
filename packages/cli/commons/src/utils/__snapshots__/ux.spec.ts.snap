// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ux formatOrgId() when shouldUseColor returns true colors the color magenta 1`] = `"[35mmyOrgId[39m"`;
